### DISCUSSION TITLE
Is `UPDATE...WHERE` atomic?

### DISCUSSION THREAD

**Discussion by mukundt** [supabase/supabase#26271] [Questions] [answered]
I have N clients that use the Supabase Python API to attempt a record update. My intended behavior is that only one of them succeeds in "claiming" that record when all N clients attempt simultaneously.

Will the below code actually guarantee that exactly one of them succeeds? Or do I need to use a Postgres function to ensure atomicity?

```
eligible_records = (
        supabase.table("records")
        .update({"id": "unique ID", "claimed": True})
        .match({"claimed": False})
        .execute()
        .data
    )
    if len(eligible_records) == 1:
        # We claimed this record!
```

**Comment by GaryAustin1** [comment, marked as answer]
The update will be an SQL update which reads and writes back the row without any other changes to that row(s) allowed until complete.
Just be careful if you use an RLS select policy involving `claimed` as that might be an issue for this operation.

**Reply by mukundt** [reply]
@GaryAustin1 thanks for the response. Just to clarify, are you saying that the `WHERE` and `UPDATE` portions of the statement happen under one single atomic transaction?

We specifically want to avoid the case where Client A and Client B both read a particular row and find that `claimed = False`, then both attempt to write `claimed = True`, and have both of those `update` statements sequentially succeed.

Or do we need to use a [Postgres function](https://www.youtube.com/watch?v=xUeuy19a-Uw) to ensure that behavior?

**Reply by GaryAustin1** [reply]
The Where is part of the Update operation.  SQL Update does not allow you to select (using the Where clause) a row that is already selected for update.  And PostgREST converts your call to an update SQL operation.  You do not need a function for that.

**Reply by mukundt** [reply]
Sounds good, thank you Gary!
